torcheeg/trainers/classifier.py

```diff
@@ -266,4 +296,349 @@ def predict_step(self,
+        
+        return loss
+        
+
+    def on_validation_epoch_end(self) -> None:
+        # log loss
+        log_dict = {"Val Total Loss" : self.val_loss.compute(),
+                    "Val CELoss" : self.predict_loss_val.compute(),
+                    "Val CLoss": self.center_loss_val.compute() }
+        # log classfication metrics
+        for i, metric_value in enumerate(self.val_metrics.values()):
+            log_dict[f"Val {self.metrics[i]}"] = metric_value.compute()
+        self.log_dict(log_dict,
+                    prog_bar=True,
+                    on_epoch=True,
+                     on_step=False,
+                     logger=True)
+        # reset the metrics
+        self.reset_metric("val")
+        str = "\n[Val] "
+        for key, value in self.trainer.logged_metrics.items():
+            if key.startswith("Val"):
+                str += f"{key}: {value:.3f} "
+        print(str + '\n')
+    
+
+
+    def test_step(self, batch: Tuple[torch.Tensor],
+                  batch_idx: int) -> torch.Tensor:
+        x, y = batch
+        # centerloss
+        feat = self.decoder(x) 
+        centerloss = self.Closs(y,feat)
+
+        # predict loss
+        y_hat = self.Fpredictor(feat)
+        pre_loss = self.ce_fn(y_hat, y)
+    
+        #Total loss 
+        loss = pre_loss +self.lammda *centerloss
+
+        self.test_loss.update(loss)
+        self.center_loss_test.update(centerloss)
+        self.predict_loss_test(pre_loss)
+        self.test_metrics.update(y_hat, y)
+        return loss
+    
+    def on_test_epoch_end(self) -> None:
+        log_dict = {"Test Total Loss" : self.test_loss.compute(),
+                    "Test CELoss" : self.predict_loss_test.compute(),
+                    "Test CLoss": self.center_loss_test.compute() }
+        # log classfication metrics
+        for i, metric_value in enumerate(self.test_metrics.values()):
+            log_dict[f"Test {self.metrics[i]}"] = metric_value.compute()
+        self.log_dict(log_dict,
+                    prog_bar=True,
+                    on_epoch=True,
+                     on_step=False,
+                     logger=True)
+        # reset the metrics
+        self.reset_metric("test")
+        str = "\n[Test] "
+        for key, value in self.trainer.logged_metrics.items():
+            if key.startswith("Test"):
+                str += f"{key}: {value:.3f} "
+        print(str + '\n')
+        
+
+        self.test_loss.reset()
+        self.test_metrics.reset()
+
+    def configure_optimizers(self):
+        parameters = list(self.model.parameters())
+        trainable_parameters = list(
+            filter(lambda p: p.requires_grad, parameters))
+        optimizer4pred = torch.optim.Adam(trainable_parameters,
+                                          lr=self.lr,
+                                          weight_decay=self.weight_decay)
+
+        optimzer4center = torch.optim.SGD(self.Closs.parameters(), lr=0.01)
```

Comment:
By the way, I'm curious as to why they opted to use Adam and SGD separately?

Reply:
Because adam is used to update models params .SGD simply caculates the mean of deapfeature code.

Reply:
Make sense, so we don't need momentum during the calculation of the mean of deapfeature.